test/pipelinefinally_test.go

```diff
@@ -0,0 +1,457 @@
+				Container: corev1.Container{Image: "alpine"},
+				Script:    "exit 0",
+			}},
+		},
+	}
+	if _, err := c.TaskClient.Create(finalTask); err != nil {
+		t.Fatalf("Failed to create final Task: %s", err)
+	}
+
+	pipeline := &v1beta1.Pipeline{
+		ObjectMeta: metav1.ObjectMeta{Name: pipelineName, Namespace: namespace},
+		Spec: v1beta1.PipelineSpec{
+			Tasks: []v1beta1.PipelineTask{{
+				Name:    "dagtask1",
+				TaskRef: &v1beta1.TaskRef{Name: taskName},
+			}, {
+				Name:    "dagtask2",
+				TaskRef: &v1beta1.TaskRef{Name: delayedTaskName},
+			}},
+			Finally: []v1beta1.PipelineTask{{
+				Name:    "finaltask1",
+				TaskRef: &v1beta1.TaskRef{Name: finalTaskName},
+			}},
+		},
+	}
+	if _, err := c.PipelineClient.Create(pipeline); err != nil {
+		t.Fatalf("Failed to create Pipeline: %s", err)
+	}
+
+	pipelineRun := &v1beta1.PipelineRun{
+		ObjectMeta: metav1.ObjectMeta{Name: pipelineRunName, Namespace: namespace},
+		Spec: v1beta1.PipelineRunSpec{
+			PipelineRef: &v1beta1.PipelineRef{Name: pipelineName},
+		},
+	}
+	if _, err := c.PipelineRunClient.Create(pipelineRun); err != nil {
+		t.Fatalf("Failed to create Pipeline Run `%s`: %s", pipelineRunName, err)
+	}
+
+	if err := WaitForPipelineRunState(c, pipelineRunName, timeout, PipelineRunFailed(pipelineRunName), "PipelineRunFailed"); err != nil {
+		t.Fatalf("Waiting for PipelineRun %s to fail: %v", pipelineRunName, err)
+	}
+
+	taskrunList, err := c.TaskRunClient.List(metav1.ListOptions{LabelSelector: "tekton.dev/pipelineRun=" + pipelineRunName})
+	if err != nil {
+		t.Fatalf("Error listing TaskRuns for PipelineRun %s: %s", pipelineRunName, err)
+	}
+
+	var dagTask1Time, dagTask2Time, finalTaskTime *metav1.Time
+	// verify non final task failed and final task executed and succeeded
+	for _, taskrunItem := range taskrunList.Items {
+		switch n := taskrunItem.Name; {
+		case strings.HasPrefix(n, pipelineRunName+"-finaltask1"):
+			if err := WaitForTaskRunState(c, n, TaskRunSucceed(n), "TaskRunSuccess"); err != nil {
+				t.Errorf("Error waiting for TaskRun to succeed: %v", err)
+			}
+			finalTaskTime = taskrunItem.Status.StartTime
+		case strings.HasPrefix(n, pipelineRunName+"-dagtask1"):
+			if !isFailed(t, n, taskrunItem.Status.Conditions) {
+				t.Fatalf("TaskRun %s for non final task should have failed", n)
+			}
+			dagTask1Time = taskrunItem.Status.CompletionTime
+		case strings.HasPrefix(n, pipelineRunName+"-dagtask2"):
+			if err := WaitForTaskRunState(c, n, TaskRunSucceed(n), "TaskRunSuccess"); err != nil {
+				t.Errorf("Error waiting for TaskRun to succeed: %v", err)
+			}
+			dagTask2Time = taskrunItem.Status.CompletionTime
+		default:
+			t.Fatalf("TaskRuns were not found for both final and dag tasks")
+		}
+	}
+	// final task should start executing after dagtask1 fails and dagtask2 is done
+	if finalTaskTime.Before(dagTask1Time) || finalTaskTime.Before(dagTask2Time) {
+		t.Fatalf("Final Tasks should start getting executed after all DAG tasks finishes")
+	}
+}
```

Comment:
my apologies but i am going to lobby a bit for slightly fewer e2e tests, if only b/c there is a lot of repetition between them. 

I like that TestPipelineLevelFinally_OneDAGTaskFailed_Failure includes a failure, that it checks the status, and that it compares the completion and start times.

I'm not sure we need these other 2 distinct test cases:
1. TestPipelineLevelFinally_Success
2. TestPipelineLevelFinally_DAGTaskSkipped_Success

TestPipelineLevelFinally_Success is covered by the example yaml test - it also checks status, times, etc., but if TestPipelineLevelFinally_OneDAGTaskFailed_Failure is doing that, it doesnt feel like there's much reason to do that again here

TestPipelineLevelFinally_DAGTaskSkipped_Success could be combined with this test by adding a skipped task into TestPipelineLevelFinally_OneDAGTaskFailed_Failure. In unit tests we definitely don't want to combine multiple test cases into one, but in end to end tests i think there is a lot of advantage since the tests are slow, the test code is long, and in the cases of these specific test cases, very repetitive

If we absolutely must have these distinct 4 tests, can they be made to share more code?

Reply:
😢  please dont apologize @bobcatfish, I understand your concern.

Dropped (1) since it is covered by the example YAML test like you mentioned and I am comfortable relying on it.

Combined (2) into `TestPipelineLevelFinally_OneDAGTaskFailed_Failure`.

Also, created common functions to avoid repetition.

Let me know if I should make any more changes to tests (e2e and unit).

Reply:
Rename `TestPipelineLevelFinally_OneDAGTaskFailed_Failure` since it has a skipped task as well now? 🤔